enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,390 @@
+type TLSConfig struct {
+	// ...
+
+	// externalCertificate provides certificate contents as a secret reference.
+	// This should be a single serving certificate, not a certificate
+	// chain. Do not include a CA certificate. The secret referenced should
+	// be present in the same namespace as that of the Route.
+	// Forbidden when `certificate` is set.
+	//
+	// +openshift:enable:FeatureSets=CustomNoUpgrade;TechPreviewNoUpgrade
+	// +optional
+	ExternalCertificate LocalObjectReference `json:"externalCertificate,omitempty" protobuf:"bytes,7,opt,name=externalCertificate"`
+}
+
+// LocalObjectReference contains enough information to let you locate the
+// referenced object inside the same namespace.
+// +structType=atomic
+type LocalObjectReference struct {
+	// name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +optional
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+}
+```
+
+_Note_: The secret is required to be created in the same namespace as that of the Route.
+The secret must be of type `kubernetes.io/tls` and the tls.key and the tls.crt key must be
+provided in the `data` (or `stringData`) field of the Secret configuration.
+
+If neither `.spec.tls.externalCertificate` or `.spec.tls.certificate` and `.spec.tls.key` are
+provided the router will serve the default generated certificates. User's will not be able to
+provide both `.spec.tls.certificate/key` and `.spec.tls.externalCertificate`. API server
+admission validation will enforce this.
+
+All valid and invalid scenarios will be depicted via the existing `RouteIngressCondition`.
+
+#### Variation
+
+N.A
+
+### Implementation Details/Notes/Constraints [optional]
+
+The router will read the secret referenced in `.spec.tls.externalCertificate` and will use
+the certificate inside to configure HAProxy if the secret is present and if the
+following pre-conditions are met:
+
+- Validations done by the router as part of [ExtendedValidateRoute()](https://github.com/openshift/router/blob/c407ebbc5d8d85daea2ef2d1ba539444a06f4d25/pkg/router/routeapihelpers/validation.go#L158) (contents of secret),
+
+  - The secret created should be in the same namespace as that of the route.
+  - The secret created is of type `kubernetes.io/tls`.
+  - Verify certificate and key (PEM encode/decode)
+  - Verify private key matches public certificate
+
+- Validations done by API server as part of [ValidateRoute()](https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go#L21),
+
+  - The router serviceaccount must have permission to read this secret particular secret.
+    - The role and rolebinding to provide this access must be provided by the user.
+  - CEL validations and o/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
+    are not specified on the route.
```

Comment:
This is like the first level validation which will be followed up by the extended validation in the router. So, I think this point can go first, before the router validation just to logically match the rel life workflow.

Reply:
Re-ordered the validations.